llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -59541,3 +59541,18 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
     return Align(1ULL << ExperimentalPrefInnermostLoopAlignment);
   return TargetLowering::getPrefLoopAlignment();
 }
+
+bool X86TargetLowering::isDesirableToCommuteWithShift(
+    const SDNode *N, CombineLevel Level) const {
+  assert((N->getOpcode() == ISD::SHL || N->getOpcode() == ISD::SRA ||
+          N->getOpcode() == ISD::SRL) &&
+         "Expected shift op");
+
+  SDValue ShiftLHS = N->getOperand(0);
+  if ((ShiftLHS.getOpcode() == ISD::SIGN_EXTEND &&
+       !(ShiftLHS->hasOneUse() && ShiftLHS.getOperand(0)->hasOneUse())) ||
+      !ShiftLHS->hasOneUse())
+    return false;
```

Comment:
Here (and everywhere else) think it would be much cleaner to just do:

```
match ShiftLHS->hasOneUse() || match(ShiftLHS, m_SExtLike(m_OneUse(m_Value(X)))
```

NB: The `m_SExtLike` needs to be added to `SDPatternMatch.h`, its just `ZEXT` + `NNEG`.

Reply:
Forget the `m_SextLike` (just do `m_SExt`) stuff, that will require more plumbing that in reasonable for this patch.

Reply:
![image](https://github.com/user-attachments/assets/2c903bb4-58c4-4455-bdad-a4cef298725b)
I feel that this makes it even less clear. What do you think?

Reply:
why do you need all the llvm::SDPatternMatch:: namespace prefixes?

Reply:
![image](https://github.com/user-attachments/assets/73049b87-7235-461c-9a03-4352d901ffbf)

Reply:
Have any other questions? Can I merge it？

Reply:
why not just add a local `using namespace llvm::SDPatternMatch` - we already do this in X86ISelLowering.cpp in several places (we get namespace clashes if we do it across the entire file).

Reply:
Just use one using at the top scope of the function.

Reply:
done~

Reply:
@goldsteinn  @RKSimon  Are there any other issues with the current MR?